### PR TITLE
fix: delete rds / Seurat artifacts when skipped in revision

### PR DESCRIPTION
**Pull request by danieljhegeman**
This commit makes sure to delete the temporary dataset artifact clone of
filetype rds that is generated during a revision IF the revised dataset
has to skip rds / Seurat.

https://app.zenhub.com/workspaces/single-cell-5e2a191dad828d52cc78b028/issues/chanzuckerberg/single-cell-data-portal/1580

### Reviewers
**Functional:** 
@MDunitz 
@ebezzi 
**Readability:** 
@metakuni 
---

## Changes
- add
- remove
- modify

### CODE REVIEW

**Comment by danieljhegeman** [comment]
Need to add corresponding test.

**Comment by danieljhegeman** [comment]
> Need to add corresponding test.

test added ✅

File: tests/unit/backend/corpora/api_server/test_revisions.py [hunk 1]
```diff
@@ -603,3 +603,31 @@ def test__with_revision_and_existing_datasets(self):
...
+        # There are two datasets -- track each id of original dataset that should now lack seurat
+        id_with_skipped_seurat = dataset_with_skipped_seurat.original_id
+
+        # Publishing collection should result in rds artifact no longer existing for dataset_with_skipped_seurat
+        self.publish_collection(self.rev_collection.id)
+        self.session.expire_all()
+
+        for dataset in self.pub_collection.db_object.datasets:
+            if dataset.id == id_with_skipped_seurat:
+                self.assertFalse(any(da.filetype == DatasetArtifactFileType.RDS for da in dataset.artifacts))
```
Comment (MDunitz) [inline tests/unit/backend/corpora/api_server/test_revisions.py:631]:
```suggestion
                self.assertFalse(any(dataset_artifact.filetype == DatasetArtifactFileType.RDS for dataset_artifact in dataset.artifacts))
```

File: tests/unit/backend/corpora/api_server/test_revisions.py [hunk 2]
```diff
@@ -603,3 +603,31 @@ def test__with_revision_and_existing_datasets(self):
...
+            else:
+                self.assertTrue(any(da.filetype == DatasetArtifactFileType.RDS for da in dataset.artifacts))
```
Comment (MDunitz) [inline tests/unit/backend/corpora/api_server/test_revisions.py:633]:
```suggestion
                self.assertTrue(any(dataset_artifact.filetype == DatasetArtifactFileType.RDS for dataset_artifact in dataset.artifacts))
```

**Review by MDunitz — APPROVED** [review]
lgtm -- is there separate work to delete any lingering dataset artifacts associated with previous revisions?

**Comment by danieljhegeman** [comment]
> lgtm -- is there separate work to delete any lingering dataset artifacts associated with previous revisions?

Great question @MDunitz -- previously it was not possible to revise a dataset that would fail seurat (since the failure would prevent updates to db), that is until the changes that allowed skipping of seurat were introduced to prod about a month and a half ago.

If any datasets were updated with files that failed seurat, but the previous version did not, then these datasets will have erroneous rds-type dataset artifact rows. 

I'll make a script to see if any of the rds-type dataset artifacts have an s3_url that does not correspond to an actual object in s3 (because all original s3 objects are deleted during revision; the rds file would be gone).

**Comment by danieljhegeman** [comment]
@MDunitz 
```
env=prod
s3_uris=`DEPLOYMENT_STAGE=$env make db/connect ARGS="-c \"select s3_uri from dataset_artifact where filetype = 'RDS'\"" | grep 's3:' | sed 's/^[ \t]*//'`
bucket=corpora-data-$env
all_files=`aws s3 ls --recursive $bucket`
echo $all_files > all_files
for s3_uri in $s3_uris; do
    uri=`echo $s3_uri | sed "s/.*$bucket\\/\\(.*\\)/\\1/"`
    exists=`echo $all_files | grep -Fo $uri`
    if [ -z $exists ]; then
        echo "No corresponding s3_file for $s3_uri"
    fi
done
```
Gives
```
s3://corpora-data-prod/46d75a1a-fc18-45a8-84c7-be16e2ad1071/local.rds
s3://corpora-data-prod/7a329557-8894-4cc9-8050-4ed57a07c4d5/local.rds
s3://corpora-data-prod/4e0802de-c631-4b99-ac66-04c4443bc10a/local.rds
s3://corpora-data-prod/ddded24c-f3fb-4742-9811-f0a23bfa2a10/local.rds
s3://corpora-data-prod/46d75a1a-fc18-45a8-84c7-be16e2ad1071/local.rds
s3://corpora-data-prod/7a329557-8894-4cc9-8050-4ed57a07c4d5/local.rds
s3://corpora-data-prod/4e0802de-c631-4b99-ac66-04c4443bc10a/local.rds
```

**Pull request merged**